linesearch.go

```diff
@@ -35,96 +39,152 @@ func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType
 	dim := len(loc.X)
 	ls.x = resize(ls.x, dim)
 	ls.dir = resize(ls.dir, dim)
+
 	ls.first = true
+	ls.nextMajor = false
+
+	copyLocation(&ls.loc, loc)
+	// Indicate that all fields of ls.loc are valid.
+	ls.eval = FuncEvaluation | GradEvaluation
+	if ls.loc.Hessian != nil {
+		ls.eval |= HessEvaluation
+	}
+
+	ls.lastStep = math.NaN()
+	ls.lastOp = NoOperation
 
-	return ls.initNextLinesearch(loc, xNext)
+	return ls.initNextLinesearch(loc.X)
 }
 
-func (ls *LinesearchMethod) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
-	if ls.iterType == SubIteration {
-		// We needed to evaluate invalid fields of Location. Now we have them
-		// and can announce MajorIteration.
-		copy(xNext, loc.X)
-		ls.evalType = NoEvaluation
-		ls.iterType = MajorIteration
-		return ls.evalType, ls.iterType, nil
-	}
+func (ls *LinesearchMethod) Iterate(loc *Location) (Operation, error) {
+	switch ls.lastOp {
+	case NoOperation:
+		// TODO(vladimir-ch): We have previously returned with an error and
+		// Init() was not called. What to do? What about ls's internal state?
+
+	case MajorIteration:
+		// We previously requested MajorIteration but since we're here, the
+		// previous location was not good enough to converge. So start the next
+		// linesearch and store the next evaluation point in loc.X.
+		return ls.initNextLinesearch(loc.X)
+
+	default:
+		// Store the result of the previously requested evaluation into ls.loc.
+		if ls.lastOp&FuncEvaluation != 0 {
+			ls.loc.F = loc.F
+		}
+		if ls.lastOp&GradEvaluation != 0 {
+			copy(ls.loc.Gradient, loc.Gradient)
+		}
+		if ls.lastOp&HessEvaluation != 0 {
+			ls.loc.Hessian.CopySym(loc.Hessian)
```

Comment:
... and this are the copies that could go away. Especially this copy will be the most wasteful. Newton will often finish the linesearch with just one iteration, then HessEvaluation will be returned, then here a copy to ls.loc and then few line below a copy of the whole location back to loc.